File: backendOrderFlow/chat-service/src/main/java/by/bsuir/chatservice/service/SupportService.java
package by.bsuir.chatservice.service;

import by.bsuir.chatservice.dto.*;
import by.bsuir.chatservice.entity.SupportTicket;
import by.bsuir.chatservice.entity.TicketMessage;
import by.bsuir.chatservice.exception.AccessDeniedException;
import by.bsuir.chatservice.exception.ResourceNotFoundException;
import by.bsuir.chatservice.repository.SupportTicketRepository;
import by.bsuir.chatservice.repository.TicketMessageRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;

@Slf4j
@Service
@RequiredArgsConstructor
public class SupportService {

	private static final String SUPPORT_EXCHANGE = "support.events";

	private final SupportTicketRepository ticketRepository;
	private final TicketMessageRepository messageRepository;
	private final RabbitTemplate rabbitTemplate;

	@Transactional
	public TicketResponse createTicket(Long companyId, Long userId, CreateTicketRequest request) {
		SupportTicket ticket = SupportTicket.builder()
				.requesterCompanyId(companyId)
				.requesterUserId(userId)
				.subject(request.subject())
				.category(request.category() != null ? request.category() : SupportTicket.TicketCategory.OTHER)
				.priority(request.priority() != null ? request.priority() : SupportTicket.TicketPriority.NORMAL)
				.build();

		ticket = ticketRepository.save(ticket);

		TicketMessage initialMessage = TicketMessage.builder()
				.ticket(ticket)
				.senderId(userId)
				.messageText(request.message())
				.isAdminReply(false)
				.build();

		messageRepository.save(initialMessage);
		log.info("Created support ticket {} for company {}", ticket.getId(), companyId);

		publishEvent("TicketCreated", Map.of(
				"ticketId", ticket.getId(),
				"companyId", companyId,
				"subject", ticket.getSubject(),
				"priority", ticket.getPriority().name()
		));

		return toResponse(ticket, 1L, toMessageResponse(initialMessage));
	}

	@Transactional(readOnly = true)
	public TicketResponse getTicket(Long ticketId, Long userId, boolean isAdmin) {
		SupportTicket ticket = ticketRepository.findById(ticketId)
				.orElseThrow(() -> new ResourceNotFoundException("SupportTicket", "id", ticketId));

		if (!isAdmin && !ticket.getRequesterUserId().equals(userId)) {
			throw new AccessDeniedException("Access denied to this ticket");
		}

		long messageCount = messageRepository.countByTicketId(ticketId);
		TicketMessage lastMessage = getLastMessage(ticketId, isAdmin);

		return toResponse(ticket, messageCount, lastMessage != null ? toMessageResponse(lastMessage) : null);
	}

	@Transactional(readOnly = true)
	public PageResponse<TicketResponse> getUserTickets(Long companyId, int page, int size) {
		Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
		Page<SupportTicket> tickets = ticketRepository.findByRequesterCompanyId(companyId, pageable);

		List<TicketResponse> responses = tickets.getContent().stream()
				.map(ticket -> {
					long count = messageRepository.countByTicketId(ticket.getId());
					TicketMessage last = getLastMessage(ticket.getId(), false);
					return toResponse(ticket, count, last != null ? toMessageResponse(last) : null);
				})
				.toList();

		return new PageResponse<>(responses, tickets.getNumber(), tickets.getSize(),
				tickets.getTotalElements(), tickets.getTotalPages(),
				tickets.isFirst(), tickets.isLast());
	}

	@Transactional(readOnly = true)
	public PageResponse<TicketResponse> getAllTickets(String status, int page, int size) {
		Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
		Page<SupportTicket> tickets;

		if (status != null && !status.isEmpty()) {
			tickets = ticketRepository.findByStatus(SupportTicket.TicketStatus.valueOf(status.toUpperCase()), pageable);
		} else {
			tickets = ticketRepository.findAllOpen(pageable);
		}

		List<TicketResponse> responses = tickets.getContent().stream()
				.map(ticket -> {
					long count = messageRepository.countByTicketId(ticket.getId());
					TicketMessage last = getLastMessage(ticket.getId(), true);
					return toResponse(ticket, count, last != null ? toMessageResponse(last) : null);
				})
				.toList();

		return new PageResponse<>(responses, tickets.getNumber(), tickets.getSize(),
				tickets.getTotalElements(), tickets.getTotalPages(),
				tickets.isFirst(), tickets.isLast());
	}

	@Transactional(readOnly = true)
	public List<TicketMessageResponse> getTicketMessages(Long ticketId, Long userId, boolean isAdmin) {
		SupportTicket ticket = ticketRepository.findById(ticketId)
				.orElseThrow(() -> new ResourceNotFoundException("SupportTicket", "id", ticketId));

		if (!isAdmin && !ticket.getRequesterUserId().equals(userId)) {
			throw new AccessDeniedException("Access denied to this ticket");
		}

		List<TicketMessage> messages;
		if (isAdmin) {
			messages = messageRepository.findByTicketIdOrderBySentAtAsc(ticketId);
		} else {
			messages = messageRepository.findPublicMessagesByTicketId(ticketId);
		}

		return messages.stream().map(this::toMessageResponse).toList();
	}

	@Transactional
	public TicketMessageResponse addMessage(Long ticketId, Long senderId, boolean isAdmin, TicketMessageRequest request) {
		SupportTicket ticket = ticketRepository.findById(ticketId)
				.orElseThrow(() -> new ResourceNotFoundException("SupportTicket", "id", ticketId));

		if (!isAdmin && !ticket.getRequesterUserId().equals(senderId)) {
			throw new AccessDeniedException("Access denied to this ticket");
		}

		if (!ticket.isOpen()) {
			throw new IllegalStateException("Cannot add message to closed ticket");
		}

		TicketMessage message = TicketMessage.builder()
				.ticket(ticket)
				.senderId(senderId)
				.messageText(request.message())
				.attachmentKey(request.attachmentKey())
				.isAdminReply(isAdmin)
				.isInternalNote(isAdmin && Boolean.TRUE.equals(request.isInternalNote()))
				.build();

		message = messageRepository.save(message);

		if (isAdmin && !Boolean.TRUE.equals(request.isInternalNote())) {
			ticket.waitForCustomer();
			ticketRepository.save(ticket);
		} else if (!isAdmin) {
			if (ticket.getStatus() == SupportTicket.TicketStatus.WAITING_FOR_CUSTOMER) {
				ticket.setStatus(SupportTicket.TicketStatus.IN_PROGRESS);
				ticketRepository.save(ticket);
			}
		}

		log.info("Added message to ticket {} by user {}", ticketId, senderId);

		publishEvent("TicketMessageAdded", Map.of(
				"ticketId", ticketId,
				"senderId", senderId,
				"isAdmin", isAdmin
		));

		return toMessageResponse(message);
	}

	@Transactional
	public TicketResponse assignTicket(Long ticketId, Long adminId) {
		SupportTicket ticket = ticketRepository.findById(ticketId)
				.orElseThrow(() -> new ResourceNotFoundException("SupportTicket", "id", ticketId));

		ticket.assignTo(adminId);
		ticket = ticketRepository.save(ticket);

		log.info("Assigned ticket {} to admin {}", ticketId, adminId);

		long messageCount = messageRepository.countByTicketId(ticketId);
		return toResponse(ticket, messageCount, null);
	}

	@Transactional
	public TicketResponse resolveTicket(Long ticketId) {
		SupportTicket ticket = ticketRepository.findById(ticketId)
				.orElseThrow(() -> new ResourceNotFoundException("SupportTicket", "id", ticketId));

		ticket.resolve();
		ticket = ticketRepository.save(ticket);

		log.info("Resolved ticket {}", ticketId);

		publishEvent("TicketResolved", Map.of("ticketId", ticketId));

		long messageCount = messageRepository.countByTicketId(ticketId);
		return toResponse(ticket, messageCount, null);
	}

	@Transactional
	public TicketResponse closeTicket(Long ticketId) {
		SupportTicket ticket = ticketRepository.findById(ticketId)
				.orElseThrow(() -> new ResourceNotFoundException("SupportTicket", "id", ticketId));

		ticket.close();
		ticket = ticketRepository.save(ticket);

		log.info("Closed ticket {}", ticketId);

		long messageCount = messageRepository.countByTicketId(ticketId);
		return toResponse(ticket, messageCount, null);
	}

	@Transactional
	public TicketResponse reopenTicket(Long ticketId) {
		SupportTicket ticket = ticketRepository.findById(ticketId)
				.orElseThrow(() -> new ResourceNotFoundException("SupportTicket", "id", ticketId));

		ticket.reopen();
		ticket = ticketRepository.save(ticket);

		log.info("Reopened ticket {}", ticketId);

		long messageCount = messageRepository.countByTicketId(ticketId);
		return toResponse(ticket, messageCount, null);
	}

	private TicketMessage getLastMessage(Long ticketId, boolean includeInternal) {
		List<TicketMessage> messages;
		if (includeInternal) {
			messages = messageRepository.findByTicketIdOrderBySentAtAsc(ticketId);
		} else {
			messages = messageRepository.findPublicMessagesByTicketId(ticketId);
		}
		return messages.isEmpty() ? null : messages.getLast();
	}

	private void publishEvent(String eventType, Map<String, Object> payload) {
		try {
			rabbitTemplate.convertAndSend(SUPPORT_EXCHANGE, eventType.toLowerCase(), payload);
		} catch (Exception e) {
			log.warn("Failed to publish event {}: {}", eventType, e.getMessage());
		}
	}

	private TicketResponse toResponse(SupportTicket ticket, long messageCount, TicketMessageResponse lastMessage) {
		return new TicketResponse(
				ticket.getId(),
				ticket.getRequesterCompanyId(),
				ticket.getRequesterUserId(),
				ticket.getSubject(),
				ticket.getStatus().name(),
				ticket.getPriority().name(),
				ticket.getCategory() != null ? ticket.getCategory().name() : null,
				ticket.getAssignedAdminId(),
				ticket.getCreatedAt(),
				ticket.getUpdatedAt(),
				ticket.getResolvedAt(),
				messageCount,
				lastMessage
		);
	}

	private TicketMessageResponse toMessageResponse(TicketMessage message) {
		return new TicketMessageResponse(
				message.getId(),
				message.getTicket().getId(),
				message.getSenderId(),
				message.getIsAdminReply(),
				message.getMessageText(),
				message.getAttachmentKey(),
				message.getIsInternalNote(),
				message.getSentAt()
		);
	}
}

File: clientOrderFlow/src/app/services/support.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import {
  ApiResponse,
  CreateTicketRequest,
  PageResponse,
  SendTicketMessageRequest,
  SupportTicket,
  TicketMessage,
  TicketStatus
} from '../models/api.models';

@Injectable({
  providedIn: 'root'
})
export class SupportService {
  private readonly API_URL = '/api/support/tickets';

  constructor(private http: HttpClient) {}

  getUserTickets(page: number = 0, size: number = 20): Observable<PageResponse<SupportTicket>> {
    const params = new HttpParams()
      .set('page', String(page))
      .set('size', String(size));

    return this.http.get<ApiResponse<PageResponse<unknown>>>(this.API_URL, { params }).pipe(
      map(response => this.mapPageResponse(response.data, page, size))
    );
  }

  createTicket(request: CreateTicketRequest): Observable<SupportTicket> {
    return this.http.post<ApiResponse<unknown>>(this.API_URL, request).pipe(
      map(response => this.mapTicket(response.data))
    );
  }

  createTicketMultipart(
    subject: string,
    message: string,
    category?: string,
    priority?: string,
    files: File[] = []
  ): Observable<SupportTicket> {
    const formData = new FormData();
    formData.append('subject', subject);
    formData.append('message', message);

    if (category) {
      formData.append('category', category);
    }
    if (priority) {
      formData.append('priority', priority);
    }
    for (const file of files) {
      formData.append('files', file, file.name);
    }

    return this.http.post<ApiResponse<unknown>>(this.API_URL, formData).pipe(
      map(response => this.mapTicket(response.data))
    );
  }

  getTicketById(ticketId: number): Observable<SupportTicket> {
    return this.http.get<ApiResponse<unknown>>(`${this.API_URL}/${ticketId}`).pipe(
      map(response => this.mapTicket(response.data))
    );
  }

  getTicketMessages(ticketId: number): Observable<TicketMessage[]> {
    return this.http.get<ApiResponse<unknown[]>>(`${this.API_URL}/${ticketId}/messages`).pipe(
      map(response => (response.data || []).map(message => this.mapMessage(message)))
    );
  }

  sendTicketMessage(ticketId: number, request: SendTicketMessageRequest): Observable<TicketMessage> {
    return this.http.post<ApiResponse<unknown>>(`${this.API_URL}/${ticketId}/messages`, request).pipe(
      map(response => this.mapMessage(response.data))
    );
  }

  sendTicketMessageMultipart(
    ticketId: number,
    message: string,
    files: File[] = [],
    isInternalNote: boolean = false
  ): Observable<TicketMessage> {
    const formData = new FormData();
    formData.append('message', message);
    formData.append('isInternalNote', String(isInternalNote));

    for (const file of files) {
      formData.append('files', file, file.name);
    }

    return this.http.post<ApiResponse<unknown>>(`${this.API_URL}/${ticketId}/messages`, formData).pipe(
      map(response => this.mapMessage(response.data))
    );
  }

  getAdminTickets(status?: TicketStatus | string, search?: string, page: number = 0, size: number = 20): Observable<PageResponse<SupportTicket>> {
    let params = new HttpParams()
      .set('page', String(page))
      .set('size', String(size));

    if (status) {
      params = params.set('status', status);
    }

    if (search) {
      params = params.set('search', search);
    }

    return this.http.get<ApiResponse<PageResponse<unknown>>>(`${this.API_URL}/admin`, { params }).pipe(
      map(response => this.mapPageResponse(response.data, page, size))
    );
  }

  assignTicket(ticketId: number): Observable<SupportTicket> {
    return this.http.post<ApiResponse<unknown>>(`${this.API_URL}/${ticketId}/assign`, {}).pipe(
      map(response => this.mapTicket(response.data))
    );
  }

  resolveTicket(ticketId: number): Observable<SupportTicket> {
    return this.http.post<ApiResponse<unknown>>(`${this.API_URL}/${ticketId}/resolve`, {}).pipe(
      map(response => this.mapTicket(response.data))
    );
  }

  closeTicket(ticketId: number): Observable<SupportTicket> {
    return this.http.post<ApiResponse<unknown>>(`${this.API_URL}/${ticketId}/close`, {}).pipe(
      map(response => this.mapTicket(response.data))
    );
  }

  reopenTicket(ticketId: number): Observable<SupportTicket> {
    return this.http.post<ApiResponse<unknown>>(`${this.API_URL}/${ticketId}/reopen`, {}).pipe(
      map(response => this.mapTicket(response.data))
    );
  }

  private mapPageResponse(data: PageResponse<unknown> | undefined, page: number, size: number): PageResponse<SupportTicket> {
    return {
      content: (data?.content || []).map(ticket => this.mapTicket(ticket)),
      totalElements: data?.totalElements || 0,
      totalPages: data?.totalPages || 0,
      size: data?.size || size,
      number: data?.number || page,
      first: data?.first ?? page === 0,
      last: data?.last ?? (data?.totalPages || 0) <= page + 1
    };
  }

  private mapTicket(item: unknown): SupportTicket {
    const raw = (item || {}) as Record<string, unknown>;
    const rawLastMessage = (raw['lastMessage'] || {}) as Record<string, unknown>;

    return {
      id: Number(raw['id'] || 0),
      ticketNumber: `TKT-${raw['id'] || '0'}`,
      companyId: Number(raw['companyId'] || raw['requesterCompanyId'] || 0),
      userId: Number(raw['userId'] || raw['requesterUserId'] || 0),
      userEmail: raw['userEmail'] ? String(raw['userEmail']) : undefined,
      subject: String(raw['subject'] || ''),
      category: String(raw['category'] || 'OTHER') as SupportTicket['category'],
      priority: this.mapPriority(raw['priority']),
      status: this.mapStatus(raw['status']),
      assignedTo: raw['assignedTo'] ? Number(raw['assignedTo']) : undefined,
      assignedAdminId: raw['assignedAdminId'] ? Number(raw['assignedAdminId']) : undefined,
      assignedAdminEmail: raw['assignedAdminEmail'] ? String(raw['assignedAdminEmail']) : undefined,
      createdAt: String(raw['createdAt'] || ''),
      updatedAt: String(raw['updatedAt'] || raw['createdAt'] || ''),
      resolvedAt: raw['resolvedAt'] ? String(raw['resolvedAt']) : undefined,
      closedAt: raw['closedAt'] ? String(raw['closedAt']) : undefined,
      lastMessageAt: rawLastMessage['createdAt'] ? String(rawLastMessage['createdAt']) : rawLastMessage['sentAt'] ? String(rawLastMessage['sentAt']) : undefined,
      lastMessage: rawLastMessage['messageText'] ? String(rawLastMessage['messageText']) : undefined,
      messageCount: Number(raw['messageCount'] || 0)
    };
  }

  private mapMessage(item: unknown): TicketMessage {
    const raw = (item || {}) as Record<string, unknown>;
    const isAdminReply = Boolean(raw['isAdminReply']);

    return {
      id: Number(raw['id'] || 0),
      ticketId: Number(raw['ticketId'] || 0),
      senderId: Number(raw['senderId'] || 0),
      senderEmail: isAdminReply ? 'admin@orderflow.local' : 'user@orderflow.local',
      senderRole: isAdminReply ? 'ADMIN' : 'USER',
      messageText: String(raw['messageText'] || ''),
      attachmentKeys: Array.isArray(raw['attachmentKeys']) ? raw['attachmentKeys'].map(value => String(value)) : [],
      attachmentKey: raw['attachmentKey'] ? String(raw['attachmentKey']) : undefined,
      isInternalNote: raw['isInternalNote'] !== undefined ? Boolean(raw['isInternalNote']) : undefined,
      createdAt: String(raw['createdAt'] || raw['sentAt'] || '')
    };
  }

  private mapStatus(status: unknown): TicketStatus {
    switch (String(status || '')) {
      case 'NEW':
        return 'OPEN';
      case 'WAITING_FOR_CUSTOMER':
        return 'WAITING_USER';
      case 'IN_PROGRESS':
        return 'IN_PROGRESS';
      case 'RESOLVED':
        return 'RESOLVED';
      case 'CLOSED':
        return 'CLOSED';
      default:
        return 'OPEN';
    }
  }

  private mapPriority(priority: unknown): SupportTicket['priority'] {
    switch (String(priority || '')) {
      case 'HIGH':
      case 'URGENT':
        return 'HIGH';
      case 'LOW':
        return 'LOW';
      default:
        return 'NORMAL';
    }
  }
}

